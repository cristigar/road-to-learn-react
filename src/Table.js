import React from 'react';
import Button from "./Button";
import PropTypes from 'prop-types';

const Table = ({list, onDismiss}) => {
  const largeColumn = {
    width: '40%',
  };

  const midColumn = {
    width: '30%',
  };

  const smallColumn = {
    width: '10%',
  };

  return (
    <div className="table">
      <div className="table-header">
        <span style={largeColumn}>Title</span>
        <span style={midColumn}>Author</span>
        <span style={smallColumn}>Number of comments</span>
        <span style={smallColumn}>Points</span>
        <span style={smallColumn}>Action</span>
      </div>
      {list.map(item =>
        <div key={item.objectID} className="table-row">
          <span style={largeColumn}>
            <a
              href={item.url}
              title="Title"
              target="_blank"
            >
              {item.title || item.story_title}
            </a>
          </span>
          <span
            style={midColumn}
            title="Author"
          >
            {item.author}
          </span>
          <span
            style={smallColumn}
            title="Number of comments"
          >
            {item.num_comments}
          </span>
          <span
            style={smallColumn}
            title="Points"
          >
            {item.points}
          </span>
          <span style={smallColumn}>
            <Button
              onClick={() => onDismiss(item.objectID)}
              className="button-inline"
            >
              Dismiss
            </Button>
          </span>
        </div>
      )}
    </div>
  );
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
};

export default Table;
