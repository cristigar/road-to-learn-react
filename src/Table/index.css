.table {
    margin: 20px 0;
}

.table-header {
    display: flex;
    line-height: 24px;
    font-size: 16px;
    padding: 0 10px;
    justify-content: space-between;
    font-weight: 700;
}

.table-empty {
    margin: 200px;
    text-align: center;
    font-size: 16px;
}

.table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    margin: 10px 0;
    padding: 10px;
    background: #ffffff;
    border: 1px solid #e3e3e3;
}

.table-header > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
    line-height: 1;
}

.table-row > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
}

.item-title-header,
.item-title {
    width: 40%;
}

.item-author-header,
.item-author {
    width: 20%;
}

.item-comments-header,
.item-comments {
    width: 15%;
}

.item-points-header,
.item-points {
    width: 15%;
}


.item-action-header,
.item-dismiss {
    width: 10%;
}

@media screen and (max-width: 750px) {
    .table-header {
        display: none;
    }

    .table-row {
        flex-direction: column;
        white-space: unset;
    }

    .table-row > span {
        width: auto;
    }

    .item-title {
        font-weight: 700;
    }

    .item-author,
    .item-comments,
    .item-points {
        color: #777;
        font-size: 14px;
    }

    .item-author:before,
    .item-comments:before,
    .item-points:before {
        content: attr(title) ": ";
        font-weight: 700;
    }

    .item-dismiss {
        display: flex;
        justify-content: flex-end;
    }
}
