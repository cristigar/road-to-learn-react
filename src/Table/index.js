import React from 'react';
import PropTypes from 'prop-types';

import './index.css';

import Button from "../Button/index";

const Table = ({list, onDismiss}) => {
  return (
    <div className="table">
      <div className="table-header">
        <span className="item-title-header">Title</span>
        <span className="item-author-header">Author</span>
        <span className="item-comments-header">Number of comments</span>
        <span className="item-points-header">Points</span>
        <span className="item-action-header">Action</span>
      </div>
      {list.map(item =>
        <div key={item.objectID} className="table-row">
          <span className="item-title">
            <a
              href={item.url}
              title="Title"
              target="_blank"
            >
              {item.title || item.story_title}
            </a>
          </span>
          <span
            title="Author"
            className="item-author"
          >
            {item.author}
          </span>
          <span
            title="Number of comments"
            className="item-comments"
          >
            {item.num_comments}
          </span>
          <span
            title="Points"
            className="item-points"
          >
            {item.points}
          </span>
          <span className="item-dismiss">
            <Button
              onClick={() => onDismiss(item.objectID)}
              className="button-inline"
            >
              Dismiss
            </Button>
          </span>
        </div>
      )}
    </div>
  );
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
};

export default Table;
