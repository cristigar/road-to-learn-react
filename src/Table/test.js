import Enzyme, { shallow } from 'enzyme/build';
import Adapter from 'enzyme-adapter-react-16';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import React from 'react';

import Table from './index';

Enzyme.configure({adapter: new Adapter()});

describe('Table', () => {
  const props = {
    list: [
      {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y', url: 'url1'},
      {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z', url: 'url2'},
    ],
    onDismiss: () => {
    },
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows 2 items in the list', () => {
    const element = shallow(<Table {...props} />);
    expect(element.find('.table-row').length).toBe(2);
  });
});
