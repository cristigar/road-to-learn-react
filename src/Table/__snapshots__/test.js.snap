// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      className="item-title-header"
    >
      Title
    </span>
    <span
      className="item-author-header"
    >
      Author
    </span>
    <span
      className="item-comments-header"
    >
      Number of comments
    </span>
    <span
      className="item-points-header"
    >
      Points
    </span>
    <span
      className="item-action-header"
    >
      Action
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="item-title"
    >
      <a
        href="url1"
        target="_blank"
        title="Title"
      >
        1
      </a>
    </span>
    <span
      className="item-author"
      title="Author"
    >
      1
    </span>
    <span
      className="item-comments"
      title="Number of comments"
    >
      1
    </span>
    <span
      className="item-points"
      title="Points"
    >
      2
    </span>
    <span
      className="item-dismiss"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="item-title"
    >
      <a
        href="url2"
        target="_blank"
        title="Title"
      >
        2
      </a>
    </span>
    <span
      className="item-author"
      title="Author"
    >
      2
    </span>
    <span
      className="item-comments"
      title="Number of comments"
    >
      1
    </span>
    <span
      className="item-points"
      title="Points"
    >
      2
    </span>
    <span
      className="item-dismiss"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
